src/midi/MidiFile.cpp

```diff
@@ -128,6 +129,40 @@ MidiFile::MidiFile(QString path, bool* ok, QStringList* log)
     printLog(log);
 }
 
+MidiFile::MidiFile(QByteArray& raw_midi, bool* ok, QStringList* log)
+{
+
+    if (!log) {
+        log = new QStringList();
+    }
+
+    _pauseTick = -1;
+    _saved = false;
+    midiTicks = 0;
+    _cursorTick = 0;
+    prot = new Protocol(this);
+    prot->addEmptyAction("File opened");
+    _tracks = new QList<MidiTrack*>();
+
+
+    for (int i = 0; i < 19; i++) {
+        channels[i] = new MidiChannel(this, i);
+    }
+
+    QDataStream* stream = new QDataStream(raw_midi);
```

Comment:
We can add constructor from QDataStream to avoid code duplication

Reply:
I agree. There was already some duplication present before. Would probably make sense to fix all of this at once.